pkg/ucp/frontend/controller/resourcegroups/util.go

```diff
@@ -59,38 +61,174 @@ func (e *InvalidError) Is(err error) bool {
+	} else if err != nil {
+		return fmt.Errorf("failed to fetch resource group %q: %w", resourceGroupID.String(), err)
+	}
+
+	return nil
+}
+
+// ValidateResourceType performs semantic validation of a proxy request against registered
+// resource types.
+//
+// Returns NotFoundError if the resource type does not exist.
+// Returns InvalidError if the request cannot be routed due to an invalid configuration.
+func ValidateResourceType(ctx context.Context, client store.StorageClient, id resources.ID, locationName string, apiVersion string) (*url.URL, error) {
+	// The strategy is to:
+	// - Look up the resource type and validate that it exists .. then
+	// - Look up the location resource, and validate that it supports the requested resource type and API version.
+
+	// We need to do both because they may not be in sync. This can be be the case if a resource type is being added or deleted.
+
+	if !isOperationResourceType(id) {
+		resourceTypeID, err := datamodel.ResourceTypeIDFromResourceID(id)
 		if err != nil {
-			// Not expected to happen.
 			return nil, err
 		}
 
-		_, err = store.GetResource[datamodel.ResourceGroup](ctx, client, resourceGroupID.String())
+		_, err = store.GetResource[datamodel.ResourceType](ctx, client, resourceTypeID.String())
 		if errors.Is(err, &store.ErrNotFound{}) {
-			return nil, &NotFoundError{Message: fmt.Sprintf("resource group %q not found", resourceGroupID.String())}
+
+			// Return the error as-is to fallback to the legacy routing behavior.
+			return nil, err
+
+			// Uncomment this when we remove the legacy routing behavior.
+			// return nil, &InvalidError{Message: fmt.Sprintf("resource type %q not found", id.Type())}
 		} else if err != nil {
-			return nil, fmt.Errorf("failed to find resource group %q: %w", resourceGroupID.String(), err)
+			return nil, fmt.Errorf("failed to fetch resource type %q: %w", id.Type(), err)
+		}
+	}
+
+	locationID, err := datamodel.ResourceProviderLocationIDFromResourceID(id, locationName)
+	if err != nil {
+		return nil, err
+	}
+
+	location, err := store.GetResource[datamodel.Location](ctx, client, locationID.String())
+	if errors.Is(err, &store.ErrNotFound{}) {
+
+		// Return the error as-is to fallback to the legacy routing behavior.
+		return nil, err
+
+		// Uncomment this when we remove the legacy routing behavior.
+		// return nil, &InvalidError{Message: fmt.Sprintf("location %q not found for resource provider %q", locationName, id.ProviderNamespace())}
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to fetch location %q: %w", locationID.String(), err)
+	}
+
+	// Check if the location supports the resource type.
+	// Resource types are case-insensitive so we have to iterate.
+	var locationResourceType *datamodel.LocationResourceTypeConfiguration
+
+	// We special-case two pseudo-resource types: "locations/operationstatuses" and "locations/operationresults".
+	// If the resource type is one of these, we can return the downstream URL directly.
```

Comment:
nit: Should we update this comment to include : If the resource type is one of these or id uses an older pattern with child resource where the last segment of the type is `operationstatuses` or `operationresults` ...

Reply:
Good point. I'll put that in the next PR.